bionic/cache.py

```diff
@@ -494,25 +494,47 @@ def _exact_metadata_url_for_query(self, query):
     def _load_metadata_from_url(self, url):
         try:
             metadata_yaml = self._fs.read_bytes(url).decode('utf8')
-            return ArtifactMetadataRecord.from_yaml(metadata_yaml)
+            record = ArtifactMetadataRecord.from_yaml(metadata_yaml)
+            record.artifact_url =\
+                Inventory._absolute_url(record.artifact_url, url)
+            return record
         except Exception as e:
             raise InternalCacheStateError.from_failure(
                 'metadata record', url, e)
 
     def _create_and_write_metadata(self, query, artifact_url):
+        metadata_url = self._exact_metadata_url_for_query(query)
+        artifact_url =\
+            Inventory._absolute_or_relative_url(artifact_url, metadata_url)
+
         metadata_record = ArtifactMetadataRecord.from_content(
             entity_name=query.entity_name,
             artifact_url=artifact_url,
             provenance=query.provenance,
         )
 
-        metadata_url = self._exact_metadata_url_for_query(query)
-
         self._fs.write_bytes(
             metadata_record.to_yaml().encode('utf8'), metadata_url)
 
         return metadata_url
 
+    # when both urls start with file://, make the artifact url relative to
+    # metadata url
+    @staticmethod
+    def _absolute_or_relative_url(artifact_url, metadata_url):
+        if not is_file_url(artifact_url) or not is_file_url(metadata_url):
+            return artifact_url
+        return os.path.relpath(artifact_url, os.path.dirname(metadata_url))
```

Comment:
Just in case you weren't aware, you can also do this with `pathlib` Paths, but whichever way you think is cleaner is fine.

Reply:
Ah I did not see `pathlib` Paths before. I tried it but it's a lot harder to use `relative_to` in `pathlib`. Even the [documentation](https://docs.python.org/3.7/library/pathlib.html#correspondence-to-tools-in-the-os-module) calls out that `os.path.relpath()` and `PurePath.relative_to()` are different enough to not consider them equivalent. So I'm going to keep it like this unless you feel strongly otherwise.